Added configuration to starlite constructor

wassafshahzad opened the pull request:
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.rst`?
- [ ] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [ ] Have you updated the reference documentation?

### Which ticket is this PR associated with ?
Closes #1120 

### What does this PR do
This PR adds the following features
- Adds a configuration object in Starlite constructor
- Adds sane defaults in AppConfig
- Updates pre-commits to fix isort issue

provinzkraut commented on starlite/app.py:
```diff
@@ -158,7 +158,7 @@ class Starlite(Router):
 
     def __init__(
```
As previously discusses in #1120, this method should be [overloaded](https://docs.python.org/3/library/typing.html#typing.overload) so type checkers will assert that it's only ever called with either an `AppConfig` instance, **or** keyword args, not both. This way we can also ensure that `route_handlers` is not `None` when using the kwarg style, which should be checked explicitly aside from that as well IMO.

peterschutt replied:
As `on-app-init` plugins can also provide routes, and routes can be registered after app creation, it is hard to definitively say that `route_handlers` would _always_ be required in the case that a config object isn't directly passed to the constructor.

I don't think that will be a very common case, and the workaround in that instance would be to pass an empty list like we have to in 1.x when instantiating an app without handlers. But it is worth considering if we need to be so strict on that `route_handlers` requirement.

wassafshahzad replied:
@provinzkraut @peterschutt I am currently trying to understand the overload decorator but unfortunately I was unable to make a signature which would make sure that __init__ is called with either AppConfig or route_handlers not both 

Best I could come up with is, I would greatly appreciate some guidance
```
    @overload
    def __init__(self, config: AppConfig) -> None:
        ...
    
    @overload
    def __init__(self, route_handlers: List["ControllerRouterHandler"]) -> None:
        ...

```

Goldziher replied:
I don't think overloads are the correct way to handle this, @starlite-api/maintainers . Since we are dealing here with Kwargs its going to be hard and a workaround will be pretty ugly. Instead I suggest the following, @wassafshahzad:

Create a static "factory" method called `from_config` on the Starlite class:

```python
class Starlite:
    def init(...):
       ...
     
    @static
    def from_config(config: "AppConfig") -> Starlite:
          return Starlite(**config.dict())
```

provinzkraut replied:
Why not a `classmethod?

Goldziher replied:
sure, a class method is also fine

wassafshahzad replied:
Thanks for all the help guys

provinzkraut commented on starlite/app.py:
```diff
@@ -200,6 +200,7 @@ def __init__(
         template_config: Optional["TemplateConfig"] = None,
         type_encoders: Optional["TypeEncodersMap"] = None,
         websocket_class: Optional[Type["WebSocket"]] = None,
+        configuration: Optional[AppConfig] = None,
```
```suggestion
        config: Optional[AppConfig] = None,
```

`config` is the canonical way we refer to configuration objects (`AppConfig`, not `AppConfiguration`)

provinzkraut commented on starlite/app.py:
```diff
@@ -277,6 +278,8 @@ def __init__(
             type_encoders: A mapping of types to callables that transform them into types supported for serialization.
             websocket_class: An optional subclass of :class:`WebSocket <starlite.connection.websocket.WebSocket>` to use for
                 websocket connections.
+            configuration: An optional instance of :class:`AppConfig` <startlite.config.AppConfig> to replace group up initial
+                configurations. If passed in addition to kwargs related to configurations, those kwargs will be ignored.
```
```suggestion
            config: An instance of :class:`AppConfig` <startlite.config.AppConfig>. If passed, configure the application with values from this object instead of kwargs.
```

peterschutt commented on starlite/app.py:
```diff
@@ -158,7 +158,7 @@ class Starlite(Router):
     def __init__(
         self,
-        route_handlers: List["ControllerRouterHandler"],
+        route_handlers: Optional[List["ControllerRouterHandler"]] = None,
```
I think you need to update your fork @wassafshahzad - the current signature for this is:

```py
route_handlers: "OptionalSequence[ControllerRouterHandler]" = None,
```

peterschutt commented on starlite/config/app.py:
```diff
@@ -47,7 +47,7 @@ class AppConfig(BaseModel):
         arbitrary_types_allowed = True
 
-    after_exception: SingleOrList[AfterExceptionHookHandler]
+    after_exception: SingleOrList[AfterExceptionHookHandler] = []
     """An application level :class:`exception hook handler <starlite.types.AfterExceptionHookHandler>` or list thereof.
```
these have also changed since you forked, e.g., `after_exception: List[AfterExceptionHookHandler]` is current typing for this.

provinzkraut commented:
@wassafshahzad What's up with this? Are you no longer working on it?

wassafshahzad commented:
> @wassafshahzad What's up with this? Are you no longer working on it?

No I am I reverted my commit and rebased from main 
Then pushed the changes causing git to think it had 0 commits
I reverted my changes since I didn't want to pollute the commit history

provinzkraut commented:
Well I was a bit confused because the PR is closed. Care to reopen it?

wassafshahzad commented:
@provinzkraut I have the PR ready but running pre-commit produces errors on lines I did not even change for example

starlite\app.py:404: error: Item "None" of "Optional[Sequence[Union[Type[Controller], Union[HTTPRouteHandler, WebsocketRouteHandler, ASGIRouteHandler], Router, Callable[..., Any]]]]" has no attribute "__iter__" (not iterable)  [union-attr]

provinzkraut commented:
Please reopen the PR and push your latest changes so this can be checked.

peterschutt commented:
> @provinzkraut I have the PR ready but running pre-commit produces errors on lines I did not even change for example
> 
> 
> 
> starlite\app.py:404: error: Item "None" of "Optional[Sequence[Union[Type[Controller], Union[HTTPRouteHandler, WebsocketRouteHandler, ASGIRouteHandler], Router, Callable[..., Any]]]]" has no attribute "__iter__" (not iterable)  [union-attr]

This is because you'd changed the types of some of the collections on the config object to be optional.
